File: backend/app/modules/news/services/sentiment.py
import uuid
from datetime import datetime, timedelta, timezone

from app.core.services.base import BaseService
from app.modules.news.entities.news import AssetSentimentSnapshot
from app.modules.news.repositories.asset_sentiment import AssetSentimentSnapshotRepository
from app.modules.news.repositories.news import NewsRepository

# Recency-weighted, confidence-shrunk aggregation. Each article's contribution
# decays with age (half-life below), and the aggregate is pulled toward
# neutral (0) when total evidence is thin, so one stale article doesn't carry
# the same signal as ten recent ones.
_HALF_LIFE_7D_DAYS = 2.0
_HALF_LIFE_30D_DAYS = 7.0
_CONFIDENCE_TARGET_WEIGHT = 5.0  # weight-equivalent of ~5 fresh articles saturates confidence
_TREND_THRESHOLD = 0.05


def _weighted_sentiment(rows: list[tuple[float, datetime]], now: datetime, half_life_days: float) -> float | None:
    if not rows:
        return None
    weight_sum = 0.0
    weighted_total = 0.0
    for score, published_at in rows:
        age_days = max(0.0, (now - published_at).total_seconds() / 86400)
        weight = 0.5 ** (age_days / half_life_days)
        weight_sum += weight
        weighted_total += weight * score
    if weight_sum == 0:
        return None
    raw_avg = weighted_total / weight_sum
    confidence = min(1.0, weight_sum / _CONFIDENCE_TARGET_WEIGHT)
    return raw_avg * confidence


class NewsSentimentService(BaseService):
    def __init__(self, news_repo: NewsRepository, sentiment_repo: AssetSentimentSnapshotRepository):
        self.news_repo = news_repo
        self.sentiment_repo = sentiment_repo

    def aggregate_asset_sentiment(self, asset_id: uuid.UUID) -> AssetSentimentSnapshot | None:
        """Recomputes the rolling 7d/30d sentiment aggregate for an asset from
        News.sentiment_score (immutable per-article scores, set at ingestion).
        Does not touch per-article sentiment itself."""
        now = datetime.now(timezone.utc)
        rows_30d = self.news_repo.list_sentiment_for_asset(asset_id, since=now - timedelta(days=30))
        if not rows_30d:
            return None

        rows_7d = [row for row in rows_30d if row[1] >= now - timedelta(days=7)]

        avg_7d = _weighted_sentiment(rows_7d, now, _HALF_LIFE_7D_DAYS)
        avg_30d = _weighted_sentiment(rows_30d, now, _HALF_LIFE_30D_DAYS)

        trend = None
        if avg_7d is not None and avg_30d is not None:
            diff = avg_7d - avg_30d
            if diff > _TREND_THRESHOLD:
                trend = "IMPROVING"
            elif diff < -_TREND_THRESHOLD:
                trend = "DETERIORATING"
            else:
                trend = "STABLE"

        snapshot = AssetSentimentSnapshot(
            asset_id=asset_id,
            snapshot_date=now.replace(hour=0, minute=0, second=0, microsecond=0),
            avg_sentiment_7d=avg_7d,
            avg_sentiment_30d=avg_30d,
            article_count_7d=len(rows_7d),
            trend=trend,
        )
        self.sentiment_repo.upsert(snapshot)
        self.sentiment_repo.session.commit()
        return snapshot

File: backend-node/src/lib/news/sentiment.ts
import { prisma } from "../../prisma";

// Recency-weighted, confidence-shrunk aggregation. Each article's contribution
// decays with age (half-life below), and the aggregate is pulled toward
// neutral (0) when total evidence is thin, so one stale article doesn't carry
// the same signal as ten recent ones. Port of NewsSentimentService /
// _weighted_sentiment.
const HALF_LIFE_7D_DAYS = 2.0;
const HALF_LIFE_30D_DAYS = 7.0;
const CONFIDENCE_TARGET_WEIGHT = 5.0; // weight-equivalent of ~5 fresh articles saturates confidence
const TREND_THRESHOLD = 0.05;

export function weightedSentiment(rows: Array<[number, Date]>, now: Date, halfLifeDays: number): number | null {
  if (rows.length === 0) return null;
  let weightSum = 0;
  let weightedTotal = 0;
  for (const [score, publishedAt] of rows) {
    const ageDays = Math.max(0, (now.getTime() - publishedAt.getTime()) / 86_400_000);
    const weight = 0.5 ** (ageDays / halfLifeDays);
    weightSum += weight;
    weightedTotal += weight * score;
  }
  if (weightSum === 0) return null;
  const rawAvg = weightedTotal / weightSum;
  const confidence = Math.min(1.0, weightSum / CONFIDENCE_TARGET_WEIGHT);
  return rawAvg * confidence;
}

export interface AssetSentimentSnapshot {
  asset_id: string;
  snapshot_date: Date;
  avg_sentiment_7d: number | null;
  avg_sentiment_30d: number | null;
  article_count_7d: number;
  trend: "IMPROVING" | "DETERIORATING" | "STABLE" | null;
}

/** Port of NewsSentimentService.aggregate_asset_sentiment. Recomputes the
 * rolling 7d/30d sentiment aggregate for an asset from news.sentiment_score
 * (immutable per-article scores, set at ingestion). Does not touch
 * per-article sentiment itself.
 *
 * Unwired this phase — Python's only call site is generate_features (the
 * evaluation/features worker), which has no Node port yet. Exported ready
 * for that future phase to wire in, same as routing.ts's resolveQuoteProvider. */
export async function aggregateAssetSentiment(assetId: string): Promise<AssetSentimentSnapshot | null> {
  const now = new Date();
  const since30d = new Date(now.getTime() - 30 * 86_400_000);
  const since7d = new Date(now.getTime() - 7 * 86_400_000);

  const rows30dRaw = await prisma.news.findMany({
    where: {
      news_assets: { some: { asset_id: assetId } },
      sentiment_score: { not: null },
      published_at: { gte: since30d },
    },
    select: { sentiment_score: true, published_at: true },
  });
  if (rows30dRaw.length === 0) return null;

  const rows30d: Array<[number, Date]> = rows30dRaw.map((r) => [r.sentiment_score as number, r.published_at as Date]);
  const rows7d = rows30d.filter(([, publishedAt]) => publishedAt >= since7d);

  const avg7d = weightedSentiment(rows7d, now, HALF_LIFE_7D_DAYS);
  const avg30d = weightedSentiment(rows30d, now, HALF_LIFE_30D_DAYS);

  let trend: AssetSentimentSnapshot["trend"] = null;
  if (avg7d !== null && avg30d !== null) {
    const diff = avg7d - avg30d;
    if (diff > TREND_THRESHOLD) trend = "IMPROVING";
    else if (diff < -TREND_THRESHOLD) trend = "DETERIORATING";
    else trend = "STABLE";
  }

  const snapshotDate = new Date(now);
  snapshotDate.setUTCHours(0, 0, 0, 0);

  const snapshot: AssetSentimentSnapshot = {
    asset_id: assetId,
    snapshot_date: snapshotDate,
    avg_sentiment_7d: avg7d,
    avg_sentiment_30d: avg30d,
    article_count_7d: rows7d.length,
    trend,
  };

  await prisma.asset_sentiment_snapshots.upsert({
    where: { asset_id_snapshot_date: { asset_id: assetId, snapshot_date: snapshotDate } },
    create: snapshot,
    update: {
      avg_sentiment_7d: snapshot.avg_sentiment_7d,
      avg_sentiment_30d: snapshot.avg_sentiment_30d,
      article_count_7d: snapshot.article_count_7d,
      trend: snapshot.trend,
    },
  });

  return snapshot;
}
